src/Compiler/Microsoft.AspNetCore.Razor.Language/src/DefaultTagHelperDescriptorBuilder.cs

```diff
@@ -135,47 +134,24 @@ public override void TagMatchingRule(Action<TagMatchingRuleDescriptorBuilder> co
 
     public override TagHelperDescriptor Build()
     {
-        var diagnostics = new HashSet<RazorDiagnostic>();
-        if (_diagnostics != null)
-        {
-            diagnostics.UnionWith(_diagnostics);
-        }
+        using var _ = HashSetPool<RazorDiagnostic>.GetPooledObject(out var diagnostics);
 
-        var allowedChildTags = Array.Empty<AllowedChildTagDescriptor>();
-        if (_allowedChildTags != null)
+        if (_diagnostics is { } existingDiagnostics)
         {
-            var allowedChildTagsSet = new HashSet<AllowedChildTagDescriptor>(AllowedChildTagDescriptorComparer.Default);
-            for (var i = 0; i < _allowedChildTags.Count; i++)
-            {
-                allowedChildTagsSet.Add(_allowedChildTags[i].Build());
-            }
-
-            allowedChildTags = allowedChildTagsSet.ToArray();
+            diagnostics.UnionWith(existingDiagnostics);
         }
 
-        var tagMatchingRules = Array.Empty<TagMatchingRuleDescriptor>();
-        if (_tagMatchingRuleBuilders != null)
-        {
-            var tagMatchingRuleSet = new HashSet<TagMatchingRuleDescriptor>(TagMatchingRuleDescriptorComparer.Default);
-            for (var i = 0; i < _tagMatchingRuleBuilders.Count; i++)
-            {
-                tagMatchingRuleSet.Add(_tagMatchingRuleBuilders[i].Build());
-            }
+        var allowedChildTags = _allowedChildTags is { } allowedChildTagBuilders
+            ? allowedChildTagBuilders.BuildAll(s_allowedChildTagSetPool)
+            : Array.Empty<AllowedChildTagDescriptor>();
 
-            tagMatchingRules = tagMatchingRuleSet.ToArray();
-        }
+        var tagMatchingRules = _tagMatchingRuleBuilders is { } tagMatchingRuleBuilders
```

Comment:
you really don't like `is not null` do you? 😉

Reply:
I don't like it if I'm trying to avoid reading a field twice. 😉